app/controllers/api/v0/comments_controller.rb

```diff
@@ -3,21 +3,43 @@ module V0
     class CommentsController < ApiController
       respond_to :json
 
-      caches_action :index,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
-
-      caches_action :show,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
+      before_action :set_cache_control_headers, only: %i[index show]
 
       def index
         article = Article.find(params[:a_id])
-        @comments = article.comments.roots.includes(:user).select(%i[id processed_html user_id ancestry])
+
+        @comments = article.comments.includes(:user).select(%i[id processed_html user_id ancestry]).arrange
+
+        set_surrogate_key_header "api_comments_index", edge_cache_keys(@comments)
       end
 
       def show
-        @comment = Comment.find(params[:id].to_i(26))
+        tree_with_root_comment = Comment.subtree_of(params[:id].to_i(26)).
+          includes(:user).
+          select(%i[id processed_html user_id ancestry]).
+          arrange
+
+        # being only one tree we know that the root comment is the first (and only) key
+        @comment = tree_with_root_comment.keys.first
+        @comments = tree_with_root_comment[@comment]
+
+        set_surrogate_key_header "api_comments_show", edge_cache_keys(tree_with_root_comment)
+      end
+
+      private
+
+      # ancestry wraps a single or multiple trees of comments into a single hash,
+      # in the case of an article comments, the hash has multiple keys (the root comments),
+      # in the case of a comment and its descendants, the hash has only one key.
+      # Either way, we need to use recursion to extract all the comment cache keys
+      # collecting both the keys of each level of root comments and their descendants
+      # NOTE: the objects are already loaded in memory by "ancestry",
+      # so no additional SQL query is performed during this extraction, avoiding N+1s
+      def edge_cache_keys(comments_trees)
+        keys = comments_trees.keys.map do |comment|
```

Comment:
You can avoid storing this in a variable here by using `flat_map`

Reply:
For some weird reason I knew about `flatMap` in JavaScript but not `flat_map` in Ruby. TIL! Thanks